internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	// heap for wrp messages, ingested by serviceQOS
+	heap []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *PriorityQueue) Enqueue(msg wrp.Message) {
+	// Check whether msg would already violate maxQueueSize.
+	if len(msg.Payload) >= pq.maxQueueSize {
+		return
+	}
+
+	// Check whether enqueuing msg would violate maxQueueSize.
+	// If it does, then determine whether to enqueue msg.
+	// Repeat until the heap no longer violates maxQueueSize.
+	for {
+		total := pq.Size() + len(msg.Payload)
+		// note, total < 0 checks for a overflow
+		if !(total > pq.maxQueueSize || total < 0) || pq.Len() == 0 {
+			break
+		}
+
+		// Determine whether msg is of lower priority than the least
+		// prioritized queued message `bottom`.
+		bottom := heap.Remove(pq, pq.Len()-1).(wrp.Message)
+		// if it is then keep bottom, otherwise discard leastMsg and queue msg (the latest)
+		if bottom.QualityOfService > msg.QualityOfService {
+			msg = bottom
+		}
+	}
+
+	heap.Push(pq, msg)
+}
+
+func (pq *PriorityQueue) Size() int {
+	var s int
+	for _, msg := range pq.heap {
+		s += len(msg.Payload)
+	}
+
+	return s
+}
+
+func (pq *PriorityQueue) IsFull() bool {
+	return len(pq.heap) == cap(pq.heap) || pq.Size() > pq.maxQueueSize
```

Comment:
forgot to update this
```suggestion
	return pq.Size() > pq.maxQueueSize
```